ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -0,0 +1,335 @@
+        totalValidSize += expectedContainerValidSize;
+      }
+    }
+
+    return totalValidSize;
+  }
+
+  /**
+   * Verify the correctness of valid data size information per log segment returned by BlobStoreStats for a given {@link TimeRange}.
+   * @param timeRange the {@link TimeRange} to be used for the verification
+   * @return the total valid data size of all log segments
+   */
+  private long verifyLogSegmentValidSize(TimeRange timeRange) throws StoreException {
+    Pair<Long, NavigableMap<String, Long>> actualLogSegmentValidSizeMap = blobStoreStats.getValidDataSizeByLogSegment(timeRange);
+    assertTrue("Valid data size collection time should be in the range",
+        timeRange.getStartTimeInMs() <= actualLogSegmentValidSizeMap.getFirst()
+            && timeRange.getEndTimeInMs() >= actualLogSegmentValidSizeMap.getFirst());
+
+    int expectedNumberOfLogSegments = 0;
+    long expectedTotalLogSegmentValidSize = 0L;
+    LogSegment logSegment = state.log.getFirstSegment();
+
+    while (logSegment != null) {
+      String logSegmentName = logSegment.getName();
+      assertTrue("Log segment: " + logSegmentName + " not found",
+          actualLogSegmentValidSizeMap.getSecond().containsKey(logSegmentName));
+
+      long expectedLogSegmentValidSize =
+          state.getValidDataSizeForLogSegment(logSegment, timeRange.getEndTimeInMs(), state.time.milliseconds());
+      long actualLogSegmentValidSize = actualLogSegmentValidSizeMap.getSecond().get(logSegmentName);
+      assertEquals("Valid data size mismatch for log segment: " + logSegmentName, expectedLogSegmentValidSize,
+          actualLogSegmentValidSize);
+
+      expectedTotalLogSegmentValidSize += expectedLogSegmentValidSize;
+      expectedNumberOfLogSegments++;
+      logSegment = state.log.getNextSegment(logSegment);
+    }
+
+    assertEquals("Mismatch in number of log segments", expectedNumberOfLogSegments,
+        actualLogSegmentValidSizeMap.getSecond().size());
+
+    Pair<Long, Long> actualTotalValidSize = blobStoreStats.getValidSize(timeRange);
+    assertTrue("Valid data size collection time should be in the range",
+        timeRange.getStartTimeInMs() <= actualTotalValidSize.getFirst()
+            && timeRange.getEndTimeInMs() >= actualLogSegmentValidSizeMap.getFirst());
+    assertEquals("Total valid data size of all log segments mismatch", expectedTotalLogSegmentValidSize,
+        actualTotalValidSize.getSecond().longValue());
+
+    return actualTotalValidSize.getSecond();
+  }
+
+  /**
+   * Go over the referenceIndex to collect valid data size information per container. The result is used for
+   * verification purposes.
+   * @param deleteAndExpirationRefTimeInMs the reference time in ms until which deletes and expiration are relevant
+   * @return a nested {@link Map} of serviceId to containerId to valid data size
+   */
+  private Map<String, Map<String, Long>> getValidSizeByContainer(long deleteAndExpirationRefTimeInMs) {
```

Comment:
you can move this to `CuratedLogIndexState` if you think it will be useful elsewhere